docs/designs/api-design.md

```diff
@@ -0,0 +1,352 @@
+the newly created program.
+
+## Get Program
+
+Get details for the Program page.
+
+`GET /program/<programId>`
+
+**Request**
+
+```json
+
+```
+
+**Response**
+
+```json
+{
+  "program": {
+    "id": "<uuid>",
+    "name": "<string>",
+    "description": "<string>",
+
+    // Internal metadata used by the backend
+    "internal": {
+        "lastUpdatedTime": "<iso-date-str>",
+        "lastUpdatedUser": "<userId>",
+    },
+
+    "projects": [
+        {
+            "id": "<uuid>",
+            "name": "<string>",
+            "description": "<string>",
+            "status": "<status>",
+            "tasks": [
+                {
+                    "id": "<uuid>",
+                    "name": "<string>",
+                    "description": "<string>",
+                    "status": "<string>",
+                    "assignee": [
+                        // List of users assigned to this task
+                        {
+                            "id": "<uuid>",
+                            "firstName": "<string>",
+                            "thumbnailColor": "<hexcode>"
+                        }
+                    ],
+                    "budget": {
+                        "amount": "<number>",
+                        "currency": "<string>",
+                    },
+                    "duration": "<string>",
+                    "estimatedStartDate": "<iso-date-str>",
+                    "estimatedEndDate": "<iso-date-str>",
+                    "actualStartDate": "<iso-date-str>",
+                    "actualEndDate": "<iso-date-str>"
+                }
+            ],
+        }
+    ]
+  },
+
+  // All information necessary for visual presentation is in this object
+  "ui": {
+      "programs": {
+        "<programId>": {
+            "projectsOrder": ["projectId1", "projectId2" ...]
+        }
+      },
+      "projects": {
+          "<projectId>": {
+            "taskOrder": ["taskId1", "taskId2" ...]
+          }
+      }
+    }
+}
+```
+
+## Update Program: Add/Update Project, Add/Update Task Overview
+
+Called to update an existing program, details about it's constituent projects, and high level details about
+individual tasks within the project. It's also used when you add a new project or a new task. It's a very
+broadly scoped API overloading a lot of functionality into one. We are using this structure for now because it
+simplifies the frontend UI logic.
+
+`POST /program/<programId>`
+
+**Request**
+Request sends an update copy of the entire structure returned by `Get Program API`. Backend will take care of splicing
+individual elements and storing them in appropriate tables.
+
+**Response**
+Returns a complete response of `Get Program API`. Makes it very easy for frontend to refresh the program page using
+the existing Program page rendering logic.
+
+**HTTP Status Code**
+
+- `409 Conflict`: This is returned when new project cannot be created because the program state stored in the database
+  has changed.
```

Comment:
Not sure about this being used so heavily for Program, Project, and Task. I understand you're doing this for simplicity (just store all the data sent in the request instead of trying to merge small updates in with existing data), but I'd at least want to see this big bang split out into individual endpoints for Program, Project, and Task.

Additionally, since this would be idempotent, it should be PUT.

As a future improvement, we can allow just sending in the changed values (especially since our UI is very heavy on the single-field-editing with the inline and save-on-unfocus)

Reply:
I agree with you. I started with a finely scoped API but then arrived at this version to get something built end-to-end. I can very clearly foresee changing this in the next few weeks. 

Also, I updated POST & PUT to be consistent with right terminology.